[PM-23713] premium badge interaction

kdenney opened the pull request:
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-23713

## 📔 Objective

This PR modifies the behavior of the premium badge in the following ways when the feature flag `pm-23713-premium-badge-opens-new-premium-upgrade-dialog` is enabled:
1. In Cloud-hosted Web, clicking on the badge now opens the same dialog used in the new navigation upgrade button, allowing the user to upgrade to premium or families immediately. After successfully upgrading, the vault is re-sync'd so that the badge no longer appears and the user can continue with what they were attempting to do.
1. In Self-hosted Web, clicking on the badge navigates to the self-hosted premium subscription page where the user can look at options and if they select an option to upgrade, they'll be sent to the cloud-hosted web vault to upgrade.
1. In Browser and Desktop, clicking on the badge now opens a new dialog with a prompt to upgrade and a button to upgrade. 
    1. When pointed to a cloud-hosted server, the upgrade button will redirect the user to the cloud web vault's subscription page.
    1. When pointed to a self-hosted server, the button will redirect to the self-hosted subscription page.

To achieve this, several changes were made in locations where the premium badge was (or should have been) used, including:
1. Ensuring that code locations that included the premium badge but also would have the parent container redirect to premium now instead use the `PremiumUpgradePromptService` in every case.
  1. This meant that I could also remove all usages of messaging service for redirecting to the premium page
  1. This also includes using the prompt service instead of the messaging service in the `has-premium.guard`

The new dialog `PremiumUpgradeDialogComponent` was created in libs/angular so that it could be used in both Browser and Desktop. This required moving the `SubscriptionPricingService` to libs/common and creating an abstraction for it and moving to libs required removing the ToastService usage from the pricing service and moving that logic to the components that use it instead.

The new dialog uses the pricing card component from libs/pricing but needed to customize some minor styles about how the card component renders. Two new customization properties were added to the card component to achieve this: `disableCardPaddingTop` and `disableCardBorder`.

A few other changes were made that require explanations and comments have been added to the files in this PR for those.

## 📸 Screenshots

Desktop dialog light mode:
<img width="1299" height="794" alt="desktop light" src="https://github.com/user-attachments/assets/d8f82719-31f7-4229-9be2-fd9c7fa67a21" />

Desktop dialog dark mode:
<img width="1303" height="799" alt="desktop dark" src="https://github.com/user-attachments/assets/f73217ea-ee15-48b4-8bba-c2bb6783ffe4" />

Browser dialog light mode:
<img width="385" height="604" alt="browser light" src="https://github.com/user-attachments/assets/f2534315-296f-4b51-a3e6-5425dc6fa479" />

Browser dialog dark mode:
<img width="385" height="605" alt="browser dark" src="https://github.com/user-attachments/assets/1941c06a-4760-4d09-a8b0-f1d4d62edcf5" />

Web vault reports page (did not use premium badge before, does now):
<img width="929" height="795" alt="web-reports-premium" src="https://github.com/user-attachments/assets/2203d8ac-10e9-4e48-b221-98e93ca898da" />

Video of the dialog being used in desktop:

https://github.com/user-attachments/assets/466270ed-862f-4565-9564-69c33388a0cf

Video of the dialog being used in browser:

https://github.com/user-attachments/assets/c2d10556-ea3e-43e1-91ee-4b7b577cfe46

Video of the premium badge interaction in cloud-hosted web vault:

https://github.com/user-attachments/assets/9cc69db8-7965-4a94-95c5-d18af8c7e818

Video of the premium badge going away after upgrading in the cloud-hosted web vault:

https://github.com/user-attachments/assets/88b4a006-6f8c-4f78-beae-be3ccb1ebd26

Video of the premium badge redirecting to the subscription page in self-host web vault:

https://github.com/user-attachments/assets/cd9228f0-6758-40b9-95d5-e2432623d6fd


## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

kdenney commented on apps/web/src/app/admin-console/organizations/policies/policy-edit-definitions/auto-confirm-policy.component.ts:
```diff
@@ -32,8 +32,8 @@ export class AutoConfirmPolicy extends BasePolicyEditDefinition {
-  private policyForm: Signal<TemplateRef<any> | undefined> = viewChild("step0");
-  private extensionButton: Signal<TemplateRef<any> | undefined> = viewChild("step1");
+  private readonly policyForm: Signal<TemplateRef<any> | undefined> = viewChild("step0");
+  private readonly extensionButton: Signal<TemplateRef<any> | undefined> = viewChild("step1");
```
I think this came from the new linter rules on signals. I didn't make the change directly so I'm pretty sure the linter did it. (same with the next file in the list)

coroiu commented on libs/angular/src/billing/components/premium-upgrade-dialog/premium-upgrade-dialog.component.ts:
```diff
@@ -0,0 +1,124 @@
+  protected loading$: Observable<boolean> = this.cardDetails$.pipe(
+    map(() => false),
+    startWith(true),
+  );
```
This will create a separate async stream where `loading$ = false` doesn't necessarily mean that anyone subscribed to `cardDetails$` directly will see the same. You'll will also always see `true` before `false` regardless of whether `cardDetails$` already has a value for you.

suggestion: solve this in the template instead https://stackoverflow.com/questions/39858526/display-loading-while-using-observable-with-async-pipe-in-template

kdenney replied:
Thanks for the suggestion. I removed the loading observable and added a loading template in the html. The dialog loads so fast I don't think it'll ever display, but it was still a good idea to add. :)

coroiu reviewed (approved):
Nice review bingo :) Platform changes look good but I saw some potential improvements in the dialog that I also commented on

kdenney commented on libs/angular/src/billing/components/premium-upgrade-dialog/premium-upgrade-dialog.component.ts:
```diff
@@ -0,0 +1,124 @@
+          title: this.i18nService.t("error"),
+          message: this.i18nService.t("unexpectedError"),
+        });
+        return of(null);
```
Good call! I added logging in https://github.com/bitwarden/clients/pull/16911/commits/447ad4920720111f26db9e47aa2c5cd79d7fbe57

kdenney commented on apps/web/src/app/vault/services/web-premium-upgrade-prompt.service.ts:
```diff
@@ -15,14 +26,43 @@ export class WebVaultPremiumUpgradePromptService implements PremiumUpgradePrompt
+      FeatureFlag.PM23713_PremiumBadgeOpensNewPremiumUpgradeDialog,
+    );
+
+    // Per conversation in PM-23713, retain the existing upgrade org flow for now, will be addressed later
```
Yes, good call. I added the ticket in https://github.com/bitwarden/clients/pull/16911/commits/447ad4920720111f26db9e47aa2c5cd79d7fbe57

coroiu reviewed (approved):
👍

Banrion reviewed (approved):
DIRT files look good!

ike-kottlowski commented on apps/web/src/app/auth/settings/emergency-access/emergency-access.component.ts:
```diff
@@ -96,15 +96,6 @@ export class EmergencyAccessComponent implements OnInit {
-      return;
-    }
-  }
-
```
❓ : Could you include in the QA notes that we need to validate this flow is unaffected. It doesn't seem like it but I would like if QA was able to double check.

kdenney commented:
> > > Changes for Tools look good, but I just noticed, that the unused message-keys aren't removed. `premiumRequired` and `premiumRequiredDesc` are just examples I found. Could you please go through and remove them from the `en/messages.json`-files
> > 
> > 
> > Those two keys are still used when the feature flag is off. I don't think there are any other keys that would be unused after these changes, either.
> 
> `premiumRequired` is no longer used in browser and desktop `premiumRequiredDesc` is no longer used in browser and desktop

@djsmith85 
Ahh, yes I was mistakenly looking at web only. While looking into this, I actually discovered that I incorrectly removed the handling of the `premiumRequired` message in desktop's app.component. When the feature flag is off, it still sends that message and without this code, the legacy premium upgrade dialog would not have appeared!

I added the message handling code back in for desktop and removed the unused message keys from browser in https://github.com/bitwarden/clients/pull/16911/commits/3373cec609b2c1733423b2cbaab18de6f3712e1e. 

Great catch and thanks for pointing this out and saving us from a bug!

cyprain-okeke commented on libs/angular/src/billing/components/premium-upgrade-dialog/premium-upgrade-dialog.component.ts:
```diff
@@ -0,0 +1,122 @@
+      map((tiers) => tiers.find((tier) => tier.id === PersonalSubscriptionPricingTierIds.Premium)),
+      map((tier) => this.mapPremiumTierToCardDetails(tier!)),
+      catchError((error: unknown) => {
+        this.toastService.showToast({
```
When an error occurs, the dialog shows a toast but remains open with no content. The user has to manually close a seemingly broken dialog. Is this the expected experience ?

kdenney replied:
Good catch! I added a line to close the dialog when this error occurs in https://github.com/bitwarden/clients/pull/16911/commits/2abc88c37dc6310d01aadb9ee3a85e065571c94f

cyprain-okeke commented on libs/angular/src/billing/components/premium-upgrade-dialog/premium-upgrade-dialog.component.ts:
```diff
@@ -0,0 +1,122 @@
+  ) {}
+
+  protected async upgrade(): Promise<void> {
+    const environment = await firstValueFrom(this.environmentService.environment$);
```
While firstValueFrom automatically unsubscribes, if the dialog is closed before these Observables emit, this could hang.

  **Recommendation:**  Add timeout or use takeUntil with component destruction
